File: surface/widgets/compose/components/widgets/primitives/Avatar.kt
// ============================================================
// Clef Surface Compose Widget — Avatar
//
// Displays a user or entity identity as initials inside a
// bordered circular surface. When no name is provided, falls
// back to a placeholder glyph. Size affects the diameter and
// text style.
//
// Adapts the avatar.widget spec: anatomy (root, image, fallback),
// states (loading, loaded, error), and connect attributes
// (data-part, data-size, data-state) to Compose rendering.
// ============================================================

package clef.surface.compose.components.widgets.primitives

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

// --------------- Helpers ---------------

private enum class AvatarSize(val sizeDp: Int, val fontSp: Int) {
    SM(32, 12),
    MD(40, 14),
    LG(56, 20),
}

private fun parseSizeParam(size: String): AvatarSize = when (size) {
    "sm" -> AvatarSize.SM
    "lg" -> AvatarSize.LG
    else -> AvatarSize.MD
}

private fun getInitials(name: String): String {
    if (name.isBlank()) return "?"
    val parts = name.trim().split("\\s+".toRegex())
    return if (parts.size == 1) {
        parts[0].first().uppercaseChar().toString()
    } else {
        "${parts.first().first().uppercaseChar()}${parts.last().first().uppercaseChar()}"
    }
}

// --------------- Component ---------------

/**
 * Avatar composable that renders user/entity identity as initials
 * inside a circular surface.
 *
 * @param name Display name used to derive initials for the fallback.
 * @param src Image source URI — triggers loaded/error states.
 * @param size Size of the avatar: "sm", "md", or "lg".
 * @param fallback Custom fallback text when no name is available.
 * @param showFallback Force the fallback to display regardless of src.
 * @param modifier Compose modifier for the root element.
 */
@Composable
fun Avatar(
    name: String = "",
    src: String? = null,
    size: String = "md",
    fallback: String? = null,
    showFallback: Boolean = false,
    modifier: Modifier = Modifier,
) {
    var state by remember { mutableStateOf(if (src != null && !showFallback) "loading" else "error") }

    LaunchedEffect(src, showFallback) {
        if (src == null || showFallback) {
            state = "error"
        } else {
            // In a non-image-loading context, transition to error to
            // show the fallback, mirroring the spec loading -> error path.
            state = "loading"
            state = "error"
        }
    }

    val config = parseSizeParam(size)
    val displayText = fallback ?: getInitials(name)

    Box(
        modifier = modifier
            .size(config.sizeDp.dp)
            .clip(CircleShape)
            .background(
                color = MaterialTheme.colorScheme.secondaryContainer,
                shape = CircleShape,
            )
            .border(
                width = 1.dp,
                color = MaterialTheme.colorScheme.outline,
                shape = CircleShape,
            ),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = displayText,
            fontWeight = FontWeight.Bold,
            fontSize = config.fontSp.sp,
            color = MaterialTheme.colorScheme.onSecondaryContainer,
            textAlign = TextAlign.Center,
        )
    }
}

File: surface/widgets/gtk/components/widgets/primitives/Avatar.ts
// ============================================================
// Clef Surface GTK Widget — Avatar
//
// Displays a user or entity identity as initials inside a
// circular Adwaita Avatar widget. When no name is provided,
// falls back to a placeholder glyph. Size affects diameter
// and text style.
//
// Adapts the avatar.widget spec: anatomy (root, image, fallback),
// states (loading, loaded, error), and connect attributes
// (data-part, data-size, data-state) to GTK4/Adwaita rendering.
// ============================================================

import Gtk from 'gi://Gtk?version=4.0';
import Adw from 'gi://Adw?version=1';

// --------------- Helpers ---------------

function parseSizeParam(size: string): number {
  switch (size) {
    case 'sm': return 32;
    case 'lg': return 56;
    default: return 40; // md
  }
}

function getInitials(name: string): string {
  if (!name.trim()) return '?';
  const parts = name.trim().split(/\s+/);
  if (parts.length === 1) {
    return parts[0][0].toUpperCase();
  }
  return parts[0][0].toUpperCase() + parts[parts.length - 1][0].toUpperCase();
}

// --------------- Props ---------------

export interface AvatarProps {
  name?: string;
  src?: string | null;
  size?: string;
  fallback?: string | null;
  showFallback?: boolean;
}

// --------------- Component ---------------

/**
 * Creates an Adwaita Avatar widget displaying user/entity identity
 * as initials inside a circular surface.
 */
export function createAvatar(props: AvatarProps = {}): Gtk.Widget {
  const {
    name = '',
    size = 'md',
    fallback = null,
    showFallback = false,
  } = props;

  const sizePx = parseSizeParam(size);
  const displayText = fallback ?? getInitials(name);

  const avatar = new Adw.Avatar({
    size: sizePx,
    text: name || displayText,
    showInitials: true,
  });

  avatar.set_tooltip_text(name || displayText);

  return avatar;
}
